### PR TITLE
Add v2 static host user resource

### PR DESCRIPTION
**Pull request by atburke**
This change adds the v2 static host user resource.

In the v1 resource, more than one SHU resource with the same login but with conflicting other fields (groups, uid/gid, etc) could apply to a node. The host user's expected configuration would be ambiguous and sensitive to the order in which the resources were created.

In the v2 resource, only one resource can exist per login (the login is now the resource name to enforce this). The SHU resource contains a list of matchers to apply different configurations to different nodes:

```yaml
kind: static_host_user
metadata:
  name: alice
spec:
  matchers:
    - node_labels:
        - name: foo
          values: ["bar"]
      groups: ["foo", "bar"]
      # sudoers, uid/gid, etc.
```

If more than one matcher applies to the node, the node can reject the host user without needing to know about any other SHU resources.

Static host users aren't in any releases yet, so no need to worry about breaking changes.

Part of #42712.

### CODE REVIEW

**Comment by zmb3** [comment]
Did we ever release the V1 resource? If not, can we make these changes without introducing a V2?

**Review by EdwardDowling — APPROVED** [review]

File: api/proto/teleport/userprovisioning/v2/statichostuser.proto [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  // this resource.
+  string node_labels_expression = 2;
+  // groups is a list of additional groups to add the user to.
+  repeated string groups = 3;
+  // sudoers is a list of sudoer entries to add.
+  repeated string sudoers = 4;
+  // uid is the new user's uid.
+  string uid = 5;
+  // gid is the new user's gid.
+  string gid = 6;
```
Comment (espadolini) [inline api/proto/teleport/userprovisioning/v2/statichostuser.proto:55]: Why are these strings?

**Review by rosstimothy — COMMENTED** [review]
Will we be deleting v1 since it's not intended to be used and was never released?

File: lib/services/statichostuser.go [hunk 2]
```diff
@@ -71,26 +72,33 @@ func ValidateStaticHostUser(u *userprovisioningpb.StaticHostUser) error {
...
+		return trace.BadParameter("missing matchers")
 	}
-	if u.Spec.NodeLabels != nil {
-		for key, value := range u.Spec.NodeLabels.Values {
-			if key == types.Wildcard && !(len(value.Values) == 1 && value.Values[0] == types.Wildcard) {
+	for _, matcher := range u.Spec.Matchers {
+		// Check if matcher can match any resources.
+		if len(matcher.NodeLabels) == 0 && len(matcher.NodeLabelsExpression) == 0 {
+			return trace.BadParameter("either NodeLabels or NodeLabelsExpression must be set")
+		}
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/services/statichostuser.go:71]: We allow both in roles, that's why I allowed both here.

**Review by rosstimothy — APPROVED** [review]

File: api/proto/teleport/userprovisioning/v2/statichostuser.proto [hunk 3]
```diff
@@ -29,30 +29,33 @@ message StaticHostUser {
...
-  string node_labels_expression = 7;
+  string node_labels_expression = 2;
+  // groups is a list of additional groups to add the user to.
+  repeated string groups = 3;
+  // sudoers is a list of sudoer entries to add.
+  repeated string sudoers = 4;
+  // uid is the new user's uid.
+  int32 uid = 5;
+  // gid is the new user's gid.
+  int32 gid = 6;
```
Comment (espadolini) [inline api/proto/teleport/userprovisioning/v2/statichostuser.proto:55]: sort of nit: other than `fixed32` and `fixed64`, protobuf numbers are variable length based on the value they contain, so there's no space saving by using `int32` over `int64`, and nothing in POSIX prevents uid_t from being 64 bits - and seeing as we're likely going to have to cast the values to a Go `int` anyway:
```suggestion
  // uid is the new user's uid.
  int64 uid = 5;
  // gid is the new user's gid.
  int64 gid = 6;
```

**Review by espadolini — APPROVED** [review]

**Pull request merged**